### PR TITLE
Updated coordinator log warning to account for busy compactors

### LINKED ISSUE #4219: Coordinator Log warn message needs improvement
**Describe the bug**
Currently the log warning message alerts when  compactors have not checked in for a given queue in a specific time interval. 

This warning level message will show up when the system is in a normal working state. 

**Versions (OS, Maven, Java, and others, as appropriate):**
 - Affected version(s) of this project: 2.1.x and newer.

**Expected behavior**
The warning log message should only be emitted when the system is not in a normal working state. 

**Additional context**
If the warn message log is updated to consider running compactions and number of jobs in the queue before emitting the message, then this would reduce the number of warn messages that don't require user action.

1. If a queue size is zero, then the following two situations are possible: 
   a. The compaction service may have been removed, along with its compactors. In 2.1, that means the tserver should not have a corresponding compaction queue defined, and the queue should be removed from the check.
   b. The queue is empty, so the compactors have backed off checking. This situation will be addressed by #4217

2. If long running compactions are keeping compactors busy, then communication is still happening so the warn message is not accurate.

### PR DESCRIPTION
**Pull request by dlmarion**
Modified the logic in CompactionCoordinator to only warn about compactors not checking in when there are idle compactors for that group. Refactored code to remove a TODO.

Fixes #4219

### CODE REVIEW

File: server/manager/src/test/java/org/apache/accumulo/manager/compaction/CompactionCoordinatorTest.java [hunk 1]
```diff
@@ -129,6 +126,13 @@ protected void startCompactionCleaner(ScheduledThreadPoolExecutor schedExecutor)
     @Override
     protected void startRunningCleaner(ScheduledThreadPoolExecutor schedExecutor) {}
 
+    @Override
+    protected void startIdleCompactionWatcher() {
+      // This is called from CompactionCoordinator.run(). Setting shutdown to true
+      // here will exit the loop in run()
```
Comment (DomGarguilo) [inline server/manager/src/test/java/org/apache/accumulo/manager/compaction/CompactionCoordinatorTest.java:132]: I think this comment needs to be updated now that this is no longer setting a boolean

Comment (dlmarion) [reply server/manager/src/test/java/org/apache/accumulo/manager/compaction/CompactionCoordinatorTest.java]: Updated comment in f5fedcd

**Review by EdColeman — APPROVED** [review]

File: server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/CompactionCoordinator.java [hunk 2]
```diff
@@ -270,35 +293,38 @@ public void run() {
...
+    } catch (InterruptedException e) {
+      LOG.warn("Interrupted waiting for shutdown latch.", e);
     }
 
     LOG.info("Shutting down");
   }
 
+  private Map<String,List<HostAndPort>> getIdleCompactors() {
+
+    Map<String,List<HostAndPort>> allCompactors = ExternalCompactionUtil.getCompactorAddrs(ctx);
```
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/CompactionCoordinator.java:309]: Changing the list to a set will be much more efficient later on the code when it keep removing things from the set.


```suggestion
    Map<String,List<HostAndPort>> allCompactors = ExternalCompactionUtil.getCompactorAddrs(ctx);
    
     Map<String,Set<HostAndPort>> allCompactors  = new HashMap<>();
    allCompactorsLists.forEach((k,v)->allCompactors.put(k, new HashSet<>(v)));
```

File: server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/CompactionCoordinator.java [hunk 3]
```diff
@@ -243,6 +243,29 @@ protected void startRunningCleaner(ScheduledThreadPoolExecutor schedExecutor) {
...
+    ThreadPools.watchNonCriticalScheduledTask(future);
+  }
+
+  private void idleCompactionWarning() {
+
+    long now = System.currentTimeMillis();
+    Map<String,List<HostAndPort>> idleCompactors = getIdleCompactors();
+    TIME_COMPACTOR_LAST_CHECKED.forEach((groupName, lastCheckTime) -> {
+      if ((now - lastCheckTime) > getMissingCompactorWarningTime()
+          && idleCompactors.containsKey(groupName.canonical())) {
```
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/CompactionCoordinator.java:259]: Could add the check for queued and probably remove the todo.  If its configured and has nothing queued we probably still do not want to log a warning.   

```suggestion
          && jobQueues.getQueuedJobs(groupName) > 0
          && idleCompactors.containsKey(groupName.canonical())) {
```

File: server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/CompactionCoordinator.java (hunk 2 above)
Comment (dlmarion) [reply server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/CompactionCoordinator.java]: Implemented in 4fe0683

Comment (dlmarion) [reply server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/CompactionCoordinator.java]: I also created an issue for 2.1 to look at changing the return type of ExternalCompactionUtil.getCompactorAddrs to benefit all uses.

**Pull request merged**